vm/src/vm.rs

```diff
@@ -227,175 +227,169 @@ impl<ENV> GVm<ENV> {
+                if m1.len() == m2.len() {
+                    if m1.is_empty() {
+                        val = Value::True;
+                    } else {
+                        // must set val to false in two instances because
+                        // its possible a previous iteration set val to true.
+                        for (k, v) in m1.iter() {
+                            if let Some(v2) = m2.get(self, k) {
+                                if self.is_equal_pair(v, v2)? == Value::False {
                                     val = Value::False;
                                     break;
+                                } else {
+                                    val = Value::True;
                                 }
+                            } else {
+                                val = Value::False;
+                                break;
                             }
                         }
                     }
                 }
-                (Value::Value(h1), Value::Value(h2)) => {
-                    let v1 = self.get_value(h1);
-                    let v2 = self.get_value(h2);
-                    val = self.is_equal_pair(v1, v2)?;
-                }
-                (val1, Value::Value(h2)) => {
-                    let v2 = self.get_value(h2);
-                    val = self.is_equal_pair(val1, v2)?;
-                }
-                (Value::Value(v1), val2) => {
-                    let v1 = self.get_value(v1);
-                    val = self.is_equal_pair(v1, val2)?;
-                }
-                (Value::Nil | Value::Undefined, Value::True) => val = Value::False,
-                (Value::Nil | Value::Undefined, Value::False) => val = Value::True,
-                (Value::True, Value::Nil | Value::Undefined) => val = Value::False,
-                (Value::False, Value::Nil | Value::Undefined) => val = Value::True,
-                (Value::Nil | Value::Undefined, Value::Nil | Value::Undefined) => val = Value::True,
-                (Value::Error(e1), Value::Error(e2)) => {
-                    let err1 = self.get_error(e1);
-                    let err2 = self.get_error(e2);
-                    if self.get_interned(err1.keyword) == self.get_interned(err2.keyword) {
-                        val = self.is_equal_pair(err1.data, err2.data)?;
-                    }
+            }
+            (Value::Value(h1), Value::Value(h2)) => {
+                let v1 = self.get_value(h1);
+                let v2 = self.get_value(h2);
+                val = self.is_equal_pair(v1, v2)?;
+            }
+            (val1, Value::Value(h2)) => {
+                let v2 = self.get_value(h2);
+                val = self.is_equal_pair(val1, v2)?;
+            }
+            (Value::Value(v1), val2) => {
+                let v1 = self.get_value(v1);
+                val = self.is_equal_pair(v1, val2)?;
+            }
+            (Value::Nil | Value::Undefined, Value::True) => val = Value::False,
```

Comment:
I am not sure undefined should be falsy.  If it sows up in this function we should probably panic, that most likely indicates a serious internal error.